flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.flink.runtime.metrics.MetricRegistry;
+import org.apache.flink.runtime.metrics.groups.JobManagerMetricGroup;
+import org.apache.flink.runtime.resourcemanager.JobLeaderIdService;
+import org.apache.flink.runtime.resourcemanager.ResourceManager;
+import org.apache.flink.runtime.resourcemanager.exceptions.ResourceManagerException;
+import org.apache.flink.runtime.resourcemanager.slotmanager.SlotManager;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.rpc.RpcService;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Kubernetes Resource Manager.
+ */
+public class KubernetesResourceManager extends ResourceManager<ResourceID> {
+	public static final String FLINK_TM_IMAGE = "FLINK_TM_IMAGE";
+	public static final String FLINK_NAMESPACE = "FLINK_NAMESPACE";
+	public static final String FLINK_TM_RESOURCE_ID = "FLINK_TM_RESOURCE_ID";
+	public static final String FLINK_CLASS_TO_RUN = "FLINK_CLASS_TO_RUN";
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesResourceManager.class);
+
+	private final Map<String, String> environment;
+
+	/**
+	 * Client to communicate with the Node manager and launch TaskExecutor processes.
+	 */
+	private KubernetesClient nodeManagerClient;
+
+	public KubernetesResourceManager(
+		RpcService rpcService,
+		String resourceManagerEndpointId,
+		ResourceID resourceId,
+		Configuration flinkConfig,
+		Map<String, String> env,
+		HighAvailabilityServices highAvailabilityServices,
+		HeartbeatServices heartbeatServices,
+		SlotManager slotManager,
+		MetricRegistry metricRegistry,
+		JobLeaderIdService jobLeaderIdService,
+		ClusterInformation clusterInformation,
+		FatalErrorHandler fatalErrorHandler,
+		JobManagerMetricGroup jobManagerMetricGroup
+	) {
+		super(
+			rpcService,
+			resourceManagerEndpointId,
+			resourceId,
+			highAvailabilityServices,
+			heartbeatServices,
+			slotManager,
+			metricRegistry,
+			jobLeaderIdService,
+			clusterInformation,
+			fatalErrorHandler,
+			jobManagerMetricGroup
+		);
+		this.environment = env;
+	}
+
+	@Override
+	protected void initialize() throws ResourceManagerException {
+		try {
+			nodeManagerClient = new DefaultKubernetesClient(environment);
+		} catch (IOException e) {
+			throw new ResourceManagerException("Error while initializing K8s client", e);
+		}
+	}
+
+	@Override
+	protected void internalDeregisterApplication(
+		ApplicationStatus finalStatus, @Nullable String optionalDiagnostics
+	) {
+		LOG.info("Shutting down and cleaning the cluster up.");
+		nodeManagerClient.stopAndCleanupCluster(null);
+	}
+
+	@Override
+	public Collection<ResourceProfile> startNewWorker(ResourceProfile resourceProfile) {
+		LOG.info("Starting a new worker.");
+		try {
+			nodeManagerClient.createClusterPod(resourceProfile);
```

Comment:
So at higher level we provide a worker with one slot only, does that strategy have a downside?

Reply:
For now, it is our basis, we consciously do the same on samza.
It's a reasonable solution because in this case, different slot threads will not compete for the CPU and memory (since task manager doesn't isolate these resources). Also recovering is easier. However, we will use a slot sharing feature and share slots between different Flink operations according to pipeline logic to get rid of high network usage between task managers.

Reply:
As for a downside, you asked, I may mention the absence of resource sharing. In the case of low job utilization, a task manager will simply stand idle without much load. 
Also, in this case, there will be no slot grouping. This feature tends to reduce network traffic by allocating slots on a single task manager. However, we will use slot sharing instead.